File: Combination_Sum_II.cpp
/*
Given a collection of candidate numbers (C) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.
Each number in C may only be used once in the combination.
Note:
All numbers (including target) will be positive integers.
Elements in a combination (a1, a2, … , ak) must be in non-descending order. (ie, a1 ≤ a2 ≤ … ≤ ak).
The solution set must not contain duplicate combinations.
For example, given candidate set 10,1,2,7,6,1,5 and target 8, 
A solution set is: 
[1, 7] 
[1, 2, 5] 
[2, 6] 
[1, 1, 6] 

Analysis:
同样是回溯法，但是需要注意相同元素的问题，每次使用了一个元素，需要将其从备选元素列表中删除。
 */
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int> > combinationSum(vector<int>& candidates, int target) {
    	vector<vector<int> > res;
    	vector<int> tmp;
    	sort(candidates.begin(), candidates.end());
    	combinate(candidates, 0, target, tmp, res);
    	return res;
    }

    void combinate(vector<int>& c, int pos, int target, vector<int> &tmp, vector<vector<int> >& res) {
    	if (target == 0 && tmp.size() > 0) {
    		res.push_back(tmp);
    		return;
    	}

    	for (int i = pos; i < c.size(); ++i) {
    		int m = c[i];
    		if (target < m)
    			break;
    		if (i > 0 && c[i] == c[i-1] && i > pos) //jump beyond a common one
    			continue;
    		tmp.push_back(m);
    		c.erase(c.begin() + i);
    		combinate(c, i, target - m, tmp, res);
    		c.insert(c.begin() + i, m);
    		tmp.pop_back();
    	}
    }
};

int main(int argc, char* argv)
{
	Solution a;
	int s[] = {1};
	vector<int> candidates (s, s + sizeof(s)/sizeof(int));
	vector<vector<int> > res = a.combinationSum(candidates, 2);
	for (int i = 0; i < res.size(); ++i) {
		for (int j = 0; j < res[i].size(); ++j) {
			cout<<res[i][j]<<" ";
		}
		cout<<endl;
	}
	return 0;
}

File: combination-sum-ii.py
# -*- coding: utf-8 -*-

"""
https://leetcode.com/problems/combination-sum-ii/

Given a collection of candidate numbers (candidates) and a target number (target),
find all unique combinations in candidates where the candidate numbers sums to target.

Each number in candidates may only be used once in the combination.

Note:
All numbers (including target) will be positive integers.
The solution set must not contain duplicate combinations.

Example 1:
Input: candidates = [10,1,2,7,6,1,5], target = 8,
A solution set is:
[
  [1, 7],
  [1, 2, 5],
  [2, 6],
  [1, 1, 6]
]

Example 2:
Input: candidates = [2,5,2,1,2], target = 5,
A solution set is:
[
  [1,2,2],
  [5]
]
"""


class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        # Sort candidates to avoid duplicate recurse
        sort_candidates = sorted(candidates)
        result = []
        # Recurse
        self.recurse(sort_candidates, target, [], result)
        return result

    def recurse(self, candidates, target, tmp_nums, result):
        # If target is 0, then tmp_nums[] is a valid combination.
        if target == 0:
            result.append(tmp_nums[:])
            return
        # If candidates is empty, then return.
        if not candidates:
            return
        # Traverse in candidates
        for i in range(len(candidates)):
            # Jump the same candidates to avoid duplicate recurse
            if i > 0 and candidates[i] == candidates[i - 1]:
                continue
            # Only use the candidate less than target
            if candidates[i] <= target:
                # The candidates[i + 1:] is next candidates
                self.recurse(candidates[i + 1:], target - candidates[i], tmp_nums + [candidates[i]], result)
            else:
                break


if __name__ == '__main__':
    s = Solution()
    print s.combinationSum2([10,1,2,7,6,1,5], 8)